pkg/controller/v1alpha1/localmodelnode/controller.go

```diff
@@ -45,13 +55,237 @@ type LocalModelNodeReconciler struct {
+		if err != nil {
+			c.Log.Error(err, "Job error", "name", jobName)
+			return err
+		}
+
+		localModelNode.Status.ModelStatus = map[string]v1alpha1api.ModelStatus{}
+		switch {
+		case job.Status.Succeeded > 0:
+			localModelNode.Status.ModelStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloaded
+		case job.Status.Failed > 0:
+			localModelNode.Status.ModelStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloadError
+		case job.Status.Ready != nil && *job.Status.Ready > 0:
+			localModelNode.Status.ModelStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloading
+		default:
+			localModelNode.Status.ModelStatus[modelInfo.ModelName] = v1alpha1api.ModelDownloadPending
+		}
+	}
+
+	if err := c.Status().Update(context.Background(), localModelNode); err != nil {
+		c.Log.Error(err, "Update local model cache status error", "name", localModelNode.Name)
+		return err
+	}
+
+	return nil
+}
+
 func (c *LocalModelNodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	c.Log.Info("Daemonset controller reconciling localmodelnode", "name", req.Name)
+	c.Log.Info("Agent reconciling LocalModelNode", "name", req.Name, "node", nodeName)
+	// Create Jobs to download models if the model is not present locally.
+	// 1. Check if LocalModelNode CR is for current node
+	localModelNode := &v1alpha1api.LocalModelNode{}
+	if err := c.Get(ctx, req.NamespacedName, localModelNode); err != nil {
+		c.Log.Error(err, "Error getting LocalModelNode", "name", req.Name)
+		return reconcile.Result{}, err
+	}
+	if nodeName != localModelNode.ObjectMeta.Name {
+		c.Log.Info("Skipping LocalModelNode because it is not for current node", "name", req.Name, "current node", nodeName, "intended node", localModelNode.ObjectMeta.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// 2. Kick off download jobs for all models in spec
+	localModelConfig, err := v1beta1.NewLocalModelConfig(c.Clientset)
+	if err != nil {
+		c.Log.Error(err, "Failed to get local model config")
+		return reconcile.Result{}, err
+	}
+	defaultJobImage = localModelConfig.DefaultJobImage
+	FSGroup = localModelConfig.FSGroup
+	if err := c.DownloadModel(ctx, localModelNode, localModelConfig.JobNamespace); err != nil {
+		c.Log.Error(err, "Model download err")
+	}
+	// Delete models from disk if they are not on the CR.
+	// 1. Scan model dir and get a list of downloaded models
+	entries, err := os.ReadDir(localModelContainerVol + "/models")
+	if err != nil {
+		c.Log.Error(err, "Failed to list model directory", "dir", localModelContainerVol+"/models")
+	}
+	localModels := map[string]struct{}{}
+	for _, entry := range entries {
+		// Models could only exist in sub dir
+		if entry.IsDir() {
+			localModels[entry.Name()] = struct{}{}
+		}
+	}
+	c.Log.Info("Found existing local models", "models", maps.Keys(localModels))
+
+	// 2. Compare with list of models from LocalModelNode CR
```

Comment:
Could you add tests for these deletion cases?

Reply:
Let's also wrap deletion code block into a function